imap_processing/idex/idex_rice_decode.py

```diff
@@ -0,0 +1,205 @@
+            # Read warmup sample
+            d1, bp = read_bits(bits, bp, n_bits)
+
+            sub_frame_data.append(d1)
+            sample_count += 1
+
+            # A 'psel' value of zero assumes that every sample in the frame is equal
+            # to the first sample in the frame. In this case, the first sample is
+            # stored as the original NBIT binary value
+            if psel == 0:
+                for _ in range(SUB_FRAME_SIZE - 1):
+                    sub_frame_data.append(d1)
+                sample_count = SUB_FRAME_SIZE
+
+        # A 'psel' value of 1 assumes that the data is not well correlated, and
+        # simply stores every sample in the frame using the original NBIT binary
+        # representation.
+        # A 'psel' value of 3 requires two uncompressed 'warm-up' samples.
+        elif (psel == 1) or ((sample_count == 1) and (psel == 3)):
+            d1, bp = read_bits(bits, bp, n_bits)
+            sub_frame_data.append(d1)
+            sample_count += 1
+
+        else:
+            # The rice parameter (k) is used to divide each re-mapped residual value
+            # into two parts. The least significant k bits of the value are called
+            # the remainder (r). The other part of the value (not included in the
+            # remainder) is called the quotient (q).
+            # The remapped quotient is unary encoded by including a number of 0 bits
+            # equal to the value of the quotient, followed by a single '1' bit.
+            q = 0
+            while read_bits(bits, bp, 1)[0] == 0:
+                bp += 1
+                q += 1
+            bp += 1
+            # If the value of the quotient is equal to or larger than 47, then a
+            # special symbol is used to denote that this particular residual value
+            # is not rice encoded, but that this special symbol is followed by the
+            # raw binary representation of the residual value using a (N_BITS+2)
+            # bit binary number. This special symbol is simply 47 zeros followed
+            # by a one.
+            if q == 47:
+                d1, bp = read_bits(bits, bp, n_bits + 2, True)
+            else:
+                if q & 0x1:
+                    q = int(-((q + 1) / 2))
+                else:
+                    q = int(q / 2)
+
+                r, bp = read_bits(bits, bp, k + 1)
+                d1 = (q << (k + 1)) + r
+
+            if psel == 2:
+                d1 = d1 + sub_frame_data[sample_count - 1]
+            elif (sample_count > 1) and (psel == 3):
+                d1 = (
+                    d1
+                    + 2 * sub_frame_data[(sample_count - 1)]
+                    - sub_frame_data[(sample_count - 2)]
+                )
+
+            if (d1 > 2**n_bits) or (d1 < -(2**n_bits)):
+                raise ValueError(
+                    f"Overflow Error while decoding subframe "
+                    f"k = {k}, q = {q}, r = {r}, d1 = {d1}\n"
```

Comment:
FYI that this can be handled directly now with simpler syntax
https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging
```suggestion
                    f"{k=}, {q=}, {r=}, {d1=}\n"
```